pkg/orchestrator/etcd_worker_test.go

```diff
@@ -0,0 +1,357 @@
+	}()
+
+	_, err := cli.Put(ctx, testEtcdKeyPrefix+"/sum", "0")
+	c.Check(err, check.IsNil)
+
+	initArray := make([]int, numValuesPerGroup)
+	jsonStr, err := json.Marshal(initArray)
+	c.Check(err, check.IsNil)
+
+	for i := 0; i < numGroups; i++ {
+		_, err := cli.Put(ctx, testEtcdKeyPrefix+"/"+strconv.Itoa(i), string(jsonStr))
+		c.Check(err, check.IsNil)
+	}
+
+	errg, ctx := errgroup.WithContext(ctx)
+	for i := 0; i < numValuesPerGroup+1; i++ {
+		finalI := i
+		errg.Go(func() error {
+			values := make([][]int, numGroups)
+			for j := range values {
+				values[j] = make([]int, numValuesPerGroup)
+			}
+
+			reactor := &simpleReactor{
+				state: nil,
+				id:    finalI,
+			}
+
+			initState := &simpleReactorState{
+				values:  values,
+				sum:     0,
+				deltas:  nil,
+				patches: nil,
+			}
+
+			cli := newClient()
+			defer func() {
+				_ = cli.Unwrap().Close()
+			}()
+
+			etcdWorker, err := NewEtcdWorker(cli, testEtcdKeyPrefix, reactor, initState)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			return errors.Trace(etcdWorker.Run(ctx, 10*time.Millisecond))
+		})
+	}
+
+	err = errg.Wait()
+	if err != nil && (errors.Cause(err) == context.DeadlineExceeded || errors.Cause(err) == context.Canceled) {
+		return
+	}
+	c.Check(err, check.IsNil)
+}
+
+type intReactorState struct {
+	val       int
+	isUpdated bool
+}
+
+func (s *intReactorState) Update(_ []byte, value []byte) {
+	var err error
+	s.val, err = strconv.Atoi(string(value))
+	if err != nil {
+		log.Panic("intReactorState", zap.Error(err))
+	}
+	s.isUpdated = true
+}
+
+func (s *intReactorState) GetPatches() []*DataPatch {
+	return []*DataPatch{}
+}
+
+type linearizabilityReactor struct {
+	state    *intReactorState
+	expected int
+}
+
+func (r *linearizabilityReactor) Tick(ctx context.Context, state ReactorState) (nextState ReactorState, err error) {
+	r.state = state.(*intReactorState)
+	if r.state.isUpdated {
+		if r.state.val != r.expected {
+			log.Panic("linearizability check failed", zap.Int("expected", r.expected), zap.Int("actual", r.state.val))
+		}
+		r.expected++
+	}
+	if r.state.val == 1999 {
+		return nil, ErrReactorFinished
+	}
+	r.state.isUpdated = false
+	return r.state, nil
+}
+
+func (s *etcdWorkerSuite) TestLinearizability(c *check.C) {
+	defer testleak.AfterTest(c)()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancel()
+
+	newClient, closer := setUpTest(c)
+	defer closer()
+
+	cli0 := newClient()
+	cli := newClient()
+	for i := 0; i < 1000; i++ {
+		_, err := cli.Put(ctx, testEtcdKeyPrefix+"/lin", strconv.Itoa(i))
+		c.Assert(err, check.IsNil)
+	}
+
+	reactor, err := NewEtcdWorker(cli0, testEtcdKeyPrefix+"/lin", &linearizabilityReactor{
+		state:    nil,
+		expected: 999,
+	}, &intReactorState{
+		val:       0,
+		isUpdated: false,
+	})
+	c.Assert(err, check.IsNil)
+	errg := &errgroup.Group{}
+	errg.Go(func() error {
+		return reactor.Run(ctx, 10*time.Millisecond)
+	})
+
+	time.Sleep(500 * time.Millisecond)
```

Comment:
Why sleep 500ms?

Reply:
The test case requires the reactor to run at least once before putting more data into Etcd.